ruby/ql/lib/codeql/ruby/dataflow/FlowSummary.qll

```diff
@@ -165,3 +167,33 @@ private class SummarizedCallableAdapter extends Impl::Public::SummarizedCallable
 }
 
 class RequiredSummaryComponentStack = Impl::Public::RequiredSummaryComponentStack;
+
+private class SummarizedCallableFromModel extends SummarizedCallable {
+  string input;
+  string output;
+  string kind;
+
+  SummarizedCallableFromModel() {
+    ModelOutput::summaryModel(input, output, kind) and
+    this = input + ";" + output + ";" + kind
```

Comment:
I would expect that this would instead be the concatenation of `package`, `type`, and `path`, since that is what identifies (calls to) the method. This also means that if a method has multiple summaries, then we will still only synthesize one method.

Reply:
> This also means that if a method has multiple summaries, then we will still only synthesize one method.

But if there are different ways to access the same method, you end up synthesizing multiple identical methods.

For example, if you had these summaries, would you like to synthesize one or two methods?
```
foo;;Member[Bar].Method[m];Argument[0];ReturnValue;taint
foo;;Member[Bar].Member[Impl].Method[m];Argument[0];ReturnValue;taint
```

From our perspective it's not well-defined if these two are "the same method".

I guess it comes down to whether or not there are adverse effects to:
1. sharing callables for unrelated methods that happen to have the same input/output/kind spec, and
2. a call having multiple call targets, due to having multiple input/output/kind specs

I can see some problems with (1) actually, as it can confuse an interprocedural analysis that has imperfect context sensitivity. The monomorphic API use assumption, for example, would not hold because these callables end up being much more polymorphic than real-world code would be. And sometimes an ad-hoc analysis does a bit of interprocedural reasoning without bothering with context sensitivity.

OK, I've convinced myself and will instead canonicalize the callables based on `package,type,path`.